internal/server/auditsink/auditsink.go

```diff
@@ -0,0 +1,204 @@
+package auditsink
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/trace"
+	"go.uber.org/zap"
+)
+
+const auditEventVersion = "v0.1"
+
+// AuditType represents what resource is being acted on.
+type AuditType string
+
+// AuditAction represents the action being taken on the resource.
+type AuditAction string
+
+const (
+	ConstraintType   AuditType = "constraint"
+	DistributionType AuditType = "distribution"
+	FlagType         AuditType = "flag"
+	RuleType         AuditType = "rule"
+	SegmentType      AuditType = "segment"
+	VariantType      AuditType = "variant"
+
+	CreateAction AuditAction = "created"
+	DeleteAction AuditAction = "deleted"
+	UpdateAction AuditAction = "updated"
+)
+
+// AuditEvent holds information that represents an audit internally.
+type AuditEvent struct {
+	Version  string         `json:"version"`
+	Metadata MetadataConfig `json:"metadata"`
+	Payload  interface{}    `json:"payload"`
+}
+
+// DecodeToAttributes provides a helper method for an AuditEvent that will return
+// a value compatible to a SpanEvent.
+func (ae AuditEvent) DecodeToAttributes() []attribute.KeyValue {
+	akv := make([]attribute.KeyValue, 0)
+
+	if ae.Version != "" {
+		akv = append(akv, attribute.KeyValue{
+			Key:   "flipt.event.version",
+			Value: attribute.StringValue(ae.Version),
+		})
+	}
+
+	if ae.Metadata.Action != "" {
+		akv = append(akv, attribute.KeyValue{
+			Key:   "flipt.event.metadata.action",
+			Value: attribute.StringValue(string(ae.Metadata.Action)),
+		})
+	}
+
+	if ae.Metadata.Type != "" {
+		akv = append(akv, attribute.KeyValue{
+			Key:   "flipt.event.metadata.type",
+			Value: attribute.StringValue(string(ae.Metadata.Type)),
+		})
+	}
+
+	if ae.Payload != nil {
+		b, err := json.Marshal(ae.Payload)
+		if err == nil {
+			akv = append(akv, attribute.KeyValue{
+				Key:   "flipt.event.payload",
+				Value: attribute.StringValue(string(b)),
+			})
+		}
+	}
+
+	return akv
+}
+
+func deserializeHelper(s string) string {
+	ss := strings.Split(s, ".")
+	return ss[len(ss)-1]
+}
+
+func (ae *AuditEvent) Valid() bool {
+	return ae.Version != "" && ae.Metadata.Action != "" && ae.Metadata.Type != "" && ae.Payload != nil
+}
+
+// decodeToAuditEvent provides helper logic for turning to value of SpanEvents to
+// an AuditEvent.
+func decodeToAuditEvent(kvs []attribute.KeyValue) (*AuditEvent, error) {
+	ae := new(AuditEvent)
+	for _, kv := range kvs {
+		terminalKey := deserializeHelper(string(kv.Key))
+
+		if terminalKey == "version" {
```

Comment:
can these been switch/case statements? will `terminalKey` ever match more than one in a loop itteration?

Reply:
@markphelps It shouldn't 🤔 . We add events to the span in one place and there only should be one of these terminal keys present for each audit event.

Reply:
yeah thats what i figured too. so using a switch would prevent checking all these if cases once one matched